go/summarize/force-graph.go

```diff
@@ -0,0 +1,190 @@
+/*
+Copyright 2024 The Vitess Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package summarize
+
+import (
+	"encoding/json"
+	"fmt"
+	"html/template"
+	"net/http"
+)
+
+type (
+	node struct {
+		ID string `json:"id"`
+	}
+
+	link struct {
+		Source    string `json:"source"`
+		SourceIdx int    `json:"source_idx"`
+		Target    string `json:"target"`
+		TargetIdx int    `json:"target_idx"`
+		Value     int    `json:"value"`
+	}
+
+	forceGraphData struct {
+		Nodes []node `json:"nodes"`
+		Links []link `json:"links"`
+	}
+)
+
+func createForceGraphData(s *Summary) forceGraphData {
+	var data forceGraphData
+
+	idxTableNode := make(map[string]int)
+	for _, table := range s.tables {
+		if len(table.JoinPredicates) > 0 {
+			data.Nodes = append(data.Nodes, node{ID: table.Table})
+			idxTableNode[table.Table] = len(data.Nodes) - 1
+		}
+	}
+	for _, join := range s.joins {
+		data.Links = append(data.Links, link{
+			Source:    join.Tbl1,
+			SourceIdx: idxTableNode[join.Tbl1],
+			Target:    join.Tbl2,
+			TargetIdx: idxTableNode[join.Tbl2],
+			Value:     join.Occurrences,
+		})
+	}
+	return data
+}
+
+func renderQueryGraph(s *Summary) {
+	data := createForceGraphData(s)
+
+	// Start the HTTP server
+	http.HandleFunc("/", func(w http.ResponseWriter, _ *http.Request) {
+		serveIndex(w, data)
+	})
+
+	port := "1010"
+	fmt.Printf("Server started at http://localhost:%s\n", port)
+	// nolint: gosec,nolintlint // this is all ran locally so no need to care about vulnerabilities around timeouts
+	if err := http.ListenAndServe(":"+port, nil); err != nil {
```

Comment:
I think we should use `127.0.0.1:<port>` here and not just `:<port>`. No need to open it to anyone outside of localhost, right?

Reply:
Fixed via https://github.com/vitessio/vt/pull/74